File: ruby/world.rb
class World
  attr_reader :width, :height

  def initialize(width, height)
    @width = width
    @height = height
    @cells = Array.new(width * height, false)
  end

  def step
    w2 = World.new(width, height)

    height.times do |y|
      width.times do |x|
        new_value = nil
        neighbor_count = count_neighbors(x, y)
        if get(x, y)
          new_value = neighbor_count == 2 || neighbor_count == 3
        else
          new_value = neighbor_count == 3
        end
        w2.set(x, y, new_value)
      end
    end

    w2
  end

  def set(x, y, val)
    idx = idx_from_coord(x, y)
    @cells[idx] = val
  end

  def get(x, y)
    idx = idx_from_coord(x, y)
    @cells[idx]
  end

  def count_neighbors(x, y)
    [
      get(x-1, y-1),
      get(x, y-1),
      get(x+1, y-1),

      get(x-1, y),
      get(x+1, y),

      get(x-1, y+1),
      get(x, y+1),
      get(x+1, y+1),
    ].count { _1 }
  end

  # idxFromCoord takes x and y coordinates and returns the index in w.cells.
  # Coordinates wrap the boundaries of the world. e.g. Given World with a
  # width of 10, then an x coordinate of -1 should be equal to 9.
  def idx_from_coord(x, y)
    x = x % width
    x += width if x < 0
    y = y % height
    y += height if y < 0

    y * width + x
  end
end

File: js/world.js
class World {
  constructor(width, height) {
    this.width = width;
    this.height = height;
    this.cells = new Array(width * height).fill(false);
  }

  step() {
    const w2 = new World(this.width, this.height);

    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        const neighborCount = this.countNeighbors(x, y);
        let newValue = null;

        if (this.get(x, y)) {
          newValue = neighborCount === 2 || neighborCount === 3;
        } else {
          newValue = neighborCount === 3;
        }

        w2.set(x, y, newValue);
      }
    }

    return w2;
  }

  set(x, y, val) {
    const idx = this.idxFromCoord(x, y);
    this.cells[idx] = val;
  }

  get(x, y) {
    const idx = this.idxFromCoord(x, y);
    return this.cells[idx];
  }

  countNeighbors(x, y) {
    const neighborOffsets = [
      [-1, -1], [0, -1], [1, -1],
      [-1, 0], [1, 0],
      [-1, 1], [0, 1], [1, 1]
    ];

    let count = 0;

    for (const [dx, dy] of neighborOffsets) {
      if (this.get(x + dx, y + dy)) {
        count++;
      }
    }

    return count;
  }

  idxFromCoord(x, y) {
    x = (x + this.width) % this.width;
    y = (y + this.height) % this.height;

    return y * this.width + x;
  }
}

module.exports = World;
